aspnetcore/mvc/advanced/custom-model-binding.md

```diff
@@ -0,0 +1,108 @@
+
+[!code-csharp[Main](custom-model-binding/sample/CustomModelBindingSample/Binders/AuthorEntityBinder.cs?name=demo)]
+
+The `AuthorEntityBinder` allows action methods to use the `Author` entity directly:
+
+[!code-csharp[Main](custom-model-binding/sample/CustomModelBindingSample/Controllers/BoundAuthorsController.cs?name=demo1&highlight=2)]
+
+In this example, since the name of the argument is not the default `authorId`, it's specified on the parameter using `ModelBinderAttribute`. Note that both the controller and action method are simplified compared to looking up the entity in the action method, since the `AppDbContext` is no longer required by the action or controller.
+
+You can apply the `ModelBinderAttribute` to individual model properties (such as on a viewmodel) or to action method parameters to specify a certain model binder or model name for just that type or action.
+
+### Implementing a ModelBinderProvider
+
+Instead of applying an attribute, you can implement [IModelBinderProvider](https://docs.microsoft.com/aspnet/core/api/microsoft.aspnetcore.mvc.modelbinding.imodelbinderprovider). This is how the built-in framework binders are implemented. Note that when you specify the type your binder operates on, you specify the type of argument it produces, not the input your binder accepts. The following binder provider would work for the `AuthorEntityBinder` implemented above, and when added to MVC's collection of providers, would eliminate the need for the `ModelBinderAttribute` on `Author`.
+
+[!code-csharp[Main](custom-model-binding/sample/CustomModelBindingSample/Binders/AuthorEntityBinderProvider.cs?highlight=17-20)]
+
+> Note:
+> The example above returns a `BinderTypeModelBinder`, which acts as a factory for model binders and provides dependency injection for them. Use it if your model binder requires services from DI.
+
+In order to configure MVC to use a custom model binder provider, you add it in `ConfigureServices`. When evaluating model binders, the collection of providers is examined in order. The first provider that returns a binder is used. Below you can see the default model binders, in order:
+
+![default model binders](custom-model-binding/images/default-model-binders.png "default model binders")
+
+Adding your provider to the end of the collection may result in it never being called. In this example, the custom provider is added to the beginning of the collection to ensure it is used for `Author` action arguments.
+
+[!code-csharp[Main](custom-model-binding/sample/CustomModelBindingSample/Startup.cs?name=callout&highlight=5-9)]
+
+## Recommendations and best practices
+
```

Comment:
Custom model binders:

- Should not attempt to set status codes or return results ..
- are most useful for eliminating repetitive code a
- Typically should not be used to convert a string into a custom type. A TypeConverter is usually a better option.

Remove "Typically, you don't need to write your own provider" and make sure it's in the intro